### PR TITLE
WIP: Readimage png + jpeg

### PR DESCRIPTION
**Pull request by r-zenine**
Hi everyone, 

This PR adds reading functions for PNG using `libpng` . It follows a discussion held in #1179.

My `C++` being a bit rusty (it's been 8 years haha). Feel free to make feedback. It will be gratefully welcomed. 
@fmassa I did my best to fix the formatting issue. Sorry about that I did not notice at the time. 

Thanks

### CODE REVIEW

**Comment by r-zenine** [comment]
Hi, 

I just noticed that CI is broken with the PR. I've looked at some of the logs. It seems to boil down to `libpng` not being found. 

Do you have any resources on how to update circleci so that i can tell it to install `libpng` ? 

Thanks

**Comment by fmassa** [comment]
Hi @r-zenine ,

We would need to install `libpng` in a few places.
We currently use both CircleCI and TravisCI (for legacy reasons).
For TravisCI, you'd need to install it in https://github.com/pytorch/vision/blob/cec7ea727ab5905ee6d7039a4f1b980124684879/.travis.yml#L44
For CircleCI, it's a bit more involved because we have builds for Linux, OSX and Windows (which are all slightly different)

For CUDA CI: https://github.com/pytorch/vision/blob/cec7ea727ab5905ee6d7039a4f1b980124684879/.circleci/config.yml.in#L120

For conda: https://github.com/pytorch/vision/blob/cec7ea727ab5905ee6d7039a4f1b980124684879/packaging/torchvision/meta.yaml#L14 (maybe using libpng from anaconda? https://github.com/pytorch/vision/blob/cec7ea727ab5905ee6d7039a4f1b980124684879/packaging/torchvision/meta.yaml#L14)

For wheels:
Somewhere like https://github.com/pytorch/vision/blob/cec7ea727ab5905ee6d7039a4f1b980124684879/packaging/build_wheel.sh#L14 maybe?

And for windows I don't actually know much... but probably somewhere in the vs folders in https://github.com/pytorch/vision/tree/master/packaging

And once we make CI pass, we still need to see how we will be seeing the story around packaging for the binaries. But let's start with small steps.

Thanks a lot for the PR!

**Comment by fmassa** [comment]
One more thing is that the `setup.py` script needs to be aware of the location of `libpng` so that it can compile everything properly, including the headers. This is something that should probably be handled in `setup.py` as well

**Comment by r-zenine** [comment]
Hi @fmassa, 

I was planning to dig deep into PIL build system this week-end. I took a first quick look at it and it seemed quite complex. Therefore, I planned to do it this week-end when I have more time. 

I will take care of the CI in the meantime. 

Could you please take a look at the code? Or do you prefer to wait until everything is done ? your call ! 

Have a good day.

File: torchvision/ops/decode_png.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    if not isinstance(input, torch.Tensor) or len(input) == 0:
+        raise ValueError("Expected a non empty 1-dimensional tensor.")
+
+    if not input.dtype == torch.uint8:
+        raise ValueError("Expected a torch.uint8 tensor.")
+    output = torch.ops.torchvision.decode_png(input)
+    return output
+
+
+class DecodePNG(nn.Module):
```
Comment (fmassa) [inline torchvision/ops/decode_png.py]: I was not thinking about adding a `nn.Module` for this operator just yet, and was thinking mostly about keeping it functional for now

File: torchvision/csrc/image/readpng.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  auto bytes = png_get_rowbytes(png_ptr, info_ptr);
+  for (decltype(height) i = 0; i < height; ++i) {
+    png_read_row(png_ptr, ptr, nullptr);
+    ptr += bytes;
+  }
+  png_destroy_read_struct(&png_ptr, &info_ptr, nullptr);
+  return tensor;
+}
+
+torch::Tensor read_png_from_file(const std::string& path) {
```
Comment (fmassa) [inline torchvision/csrc/image/readpng.cpp]: I was thinking that for simplicity, we could have this function be implemented directly in python, via something like
```python
def read_png(path):
    size = os.path.getsize(path)
    data = torch.from_file(path, dtype=torch.uint8, size=size)
    return decode_fpn(data)
```
and, if needed in the future, we could add it afterwards. Thoughts?

File: torchvision/csrc/image/readpng.h [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#ifndef READPNG_H
```
Comment (fmassa) [inline torchvision/csrc/image/readpng.h]: Can we use `#pragma once` instead?

File: torchvision/csrc/image/readpng.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+#define READPNG_H
+
+#include <torch/torch.h>
```
Comment (fmassa) [inline torchvision/csrc/image/readpng.h]: I was also thinking that we should maybe move this folder inside `csrc/cpu`, because the decoding happens on the CPU, and we could potentially add GPU decoding in the future. Thoughts?

File: torchvision/ops/decode_png.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import torch
```
Comment (fmassa) [inline torchvision/ops/decode_png.py]: I was wondering if those files should be moved to `torchvision/io/image.py` instead.
They are decoding operators, so not exactly IO operators, but we do have some video decoding there.

Thoughts?

File: torchvision/csrc/image/readpng.h (hunk 3 above)
Comment (r-zenine) [reply torchvision/csrc/image/readpng.h]: Ok, I'll make the change.

File: torchvision/ops/decode_png.py (hunk 1 above)
Comment (r-zenine) [reply torchvision/ops/decode_png.py]: As I am discovering the project. I mostly mimicked the implementation of the other operators. 

If you think it's too soon to have an `nn.Module` for this. sure, we can remove it. 

Out of curiosity and to help me understand better, why this might be premature ? Do you want to wait until we have more feedback from users ?

File: torchvision/csrc/image/readpng.cpp (hunk 2 above)
Comment (r-zenine) [reply torchvision/csrc/image/readpng.cpp]: Hi, 

You make a good point. I think I got carried away. 
It makes sense to reduce the C++ surface API. `torchvision` is mostly used through `python` anyway.  

I'll do it.

File: torchvision/ops/decode_png.py (hunk 5 above)
Comment (r-zenine) [reply torchvision/ops/decode_png.py]: Hi, 
It makes sense. My initial understanding was that ops was for `python` to `C++` bindings. I did not go through the whole codebase. I put that in there for that reason. 

I'll move the code

File: torchvision/csrc/image/readpng.h (hunk 4 above)
Comment (r-zenine) [reply torchvision/csrc/image/readpng.h]: You are right, I'll do it.

**Comment by r-zenine** [comment]
@fmassa Thanks for the review. 
I'll try to fix everything in the coming days. I'll message you when everything is updated. 
Have a good evening !

File: torchvision/csrc/image/readpng.h (hunk 4 above)
Comment (r-zenine) [reply torchvision/csrc/image/readpng.h]: I've moved the code. 

I just left an  `image.h` header file in `csrc` to serve as an entry point for the image features. 

I moved the actual implementation in `csrc/cpu` and remove the namespaces. I looked at the code base a bit and I have seen that namespaces are not used in the code base. I removed it to keep the consistency of the code.
 
Thanks

File: torchvision/ops/decode_png.py (hunk 1 above)
Comment (fmassa) [reply torchvision/ops/decode_png.py]: It is not unreasonable to have a `nn.Module` for it. Users could then write their transform pipeline as follows:
```
transforms = Compose([
    DecodePNG(),
    Resize(224),
    RandomHorizontalFlip(0.5)
])
```
which would be nice. But on another hand, chaining `Compose` functions is limiting in a few ways (see https://github.com/pytorch/vision/issues/1406 and related issues for some discussion), so I would prefer to start with just the functional interface.
My reasoning is that it's easier to add a function than to remove it once it's been added (because it would break backwards-compatibility).

Does this make sense?

**Comment by fmassa** [comment]
Hi @r-zenine 

Sounds good, looking forward to the updated version!

**Comment by r-zenine** [comment]
Hi @fmassa, 

I managed to fix one part of the CI ( Travis, linux_wheel, macos_conda ). I spent a day trying various tricks without success for the remaining pipelines. I have no experience with windows nor with anaconda packaging. I might not be able to do more. Can anyone of the team help?

Also, I looked into how we could ship the `libpng` with the package. My understanding is that adding this line below will add a `-lpng` flag when linking `vision` `C++` code. I should be enough to build `vision` from source if `libpng` is installed in a standard location ( At least for Linux).

https://github.com/r-zenine/vision/blob/fd980232b76216364a5fc5d846be073dee907b59/setup.py#L146

The questions that remain are : 

- Do we support non-standard installs of libpng via build options on Linux? 
- How does it work for macOS and Windows?

**Comment by vadimkantorov** [comment]
If extra care isn't taken to disable GCC mulitlib / multiarch and standard location OS headers, we can again run into the old linking problem when building from source: (my old investigation here:  https://github.com/torch/image/issues/171#issuecomment-255909537 )

**Comment by r-zenine** [comment]
Hi @vadimkantorov, 

My understanding is that `cmake` allows configuring a custom location for libraries. 
It seems to me that it is the same as when we want to build `torchvision` from source against a specific build of `libtorch`. I the later situation, we can configure a custom `libtorch` location via an environment variable `Torch_DIR`. I think it's possible to do the same with `libpng`. If I am not mistaken the variable should be `PNG_DIR`. 

Would this strategy address the issue you raise? Am I missing something? 

Thanks

**Comment by vadimkantorov** [comment]
@r-zenine In that old investigation there were two problems: 1) CMake passing library filenames instead of full paths even if a custom library location was specified (maybe this is no longer relevant), 2) at linking time GCC preferring system libraries over custom-built ones (including conda ones) because of multilib / multiarch. Don't remember how binary loading was mixing with this.

Differently from many PyTorch dependencies, libjpg/libpng are often present in system OS locations.

Concretely, I'm proposing having some test to confirm that build and OS binary loader picks the right libraries in presence of multilib/multiarch system library versions.

I also brought this up in https://github.com/pytorch/pytorch/issues/13188

**Comment by fmassa** [comment]
@r-zenine I currently don't have a good solution to the problem, and I'll only have more time to look into it in January unfortunately :-/

**Comment by r-zenine** [comment]
Hi @fmassa, 

Thanks already for your time. Here is what I propose: 
I can start working on the `jpeg` code in the meantime. 
When I am done, we can have a look at the CI pipeline for both PR's  and merge all the work at the same time. 

Thoughts?

**Comment by r-zenine** [comment]
Hi @fmassa, 

I just added the `jpeg` features to the same `PR`. I also tried to fix the `conda` pipeline for `linux` without success so far. 

And happy new year everyone.

**Comment by r-zenine** [comment]
Hi @fmassa , 

Any update ? 

Thanks

**Comment by fmassa** [comment]
Hi @r-zenine 

Happy new year!

Sorry for the delay in replying, after holidays I busy with a number of things. I'll have a look at your PR and get back to you

**Comment by fmassa** [comment]
I spent some time today looking into this, but I currently don't have a good solution for ensuring that users are able to compile torchvision with libpng and libjpeg.
I'll keep looking at it early next week (I'm packed with a number of things to finish for now), sorry for not having the answer yet!

One thing I noticed about the JPEG implementation is that you are seem to be using `libjpeg-turbo` -specific API, which means that if the user has only libjpeg (but not libjpeg-turbo) installed the code wouldn't compile. IIRC, `libjpeg-turbo` comes with an API that also mimics `libjpeg`, so I would think that it would be preferable as it would enable more users to install torchvision right away.

But no need to modify anything for now, as we are still figuring out how to ensure that we ship with the correct libraries.

File: torchvision/io/image.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    Returns:
+        output (Tensor[image_width, image_height, 3])
+    """
+    if not os.path.isfile(path):
+        raise ValueError("Excepted a valid file path.")
+
+    size = os.path.getsize(path)
+    if size == 0:
+        raise ValueError("Excepted a non empty file.")
```
Comment (vadimkantorov) [inline torchvision/io/image.py]: typo? "Expected"? same typo above

Comment (r-zenine) [reply torchvision/io/image.py]: Yes typo, good catch. I will fix it

**Comment by cpuhrsch** [comment]
Hello @r-zenine,

Francisco is getting flooded with other work right now so he asked me to look into this. With regards to your CI issues, could I ask you to try creating a submodule folder "third_party" similar to [pytorch](https://github.com/pytorch/pytorch/tree/master/third_party). We can then statically build this code into our binaries. Since these two libraries are so widely used I imagine we'll get good support for various environments out of the box. This will solve our deployment and CI issues (for now) until we come up with a better solution than submodules.

Thanks,
Christian

**Comment by r-zenine** [comment]
Hi @cpuhrsch , 

Sorry for the late answer. It's been three intense days on my side.
I will take a look this week-end and try to do that and ping you when I am done. 

Thanks

**Comment by r-zenine** [comment]
Hi @cpuhrsch, 

I looked at it on the week-end, here are the updates : 
- I added `zlib`, `libjpeg-turbo` and `libpng` as submodules.
- I updated the `setup.py` to build these before building the `C++` extension.

Tests on the `linux_conda` pipeline are failing, I need to investigate why. 
`simd` support on `macOs` is deactivated because of an obscure `nasm` failure. I need to take a look later. 

I'll keep looking into it tomorrow morning.

**Comment by cpuhrsch** [comment]
Great! Thanks @r-zenine!

File: setup.py [hunk 7]
```diff
@@ -76,16 +75,46 @@ def write_version_file():
 pillow_req = 'pillow-simd' if get_dist('pillow-simd') is not None else 'pillow'
 requirements.append(pillow_req + pillow_ver)
 
+third_party_libraries = ['png', 'turbojpeg']
```
Comment (cpuhrsch) [inline setup.py]: Is there a reason you chose this over adding it to the [existing CMakeLists.txt](https://github.com/pytorch/vision/blob/master/CMakeLists.txt) file?

File: setup.py [hunk 8]
```diff
@@ -196,29 +227,32 @@ def run(self):
         # It's an old-style class in Python 2.7...
         distutils.command.clean.clean.run(self)
 
+def build_ext_with_dependencies(self):
+    build_dependencies()
+    return BuildExtension.with_options(no_python_abi_suffix=True)(self)
```
Comment (cpuhrsch) [inline setup.py]: We recently landed Ninja support which spawns a lot of threads due to parallel build. The CI might fail until we introduced a flag to control the number of threads. You can disable ninja by explicitly passing "use_ninja=False" to with_options.

File: setup.py [hunk 9]
```diff
@@ -76,16 +75,46 @@ def write_version_file():
...
+        os.system("cmake3.exe -G='Visual Studio 15 2017' {} .".format(args) )
+        os.system("cmake3.exe --build .")
+    else: 
+        os.system("cmake --clean .")
+        os.system("cmake {} .".format(args) )
+        os.system("cmake --build .")
+    os.chdir(cwd)
+
+
+def build_dependencies():
```
Comment (cpuhrsch) [inline setup.py]: [This](https://github.com/pytorch/pytorch/blob/master/setup.py#L286) is how PyTorch set this up in case you didn't see it. There might be some patterns that could be borrowed from this.

File: test/test_image.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+            if ext == img_ext:
+                yield os.path.join(root, fl)
+
+
+class ImageTester(unittest.TestCase):
+    def test_read_png(self):
+        for img_path in get_images(IMAGE_DIR, "png"):
+            img_pil = torch.from_numpy(np.array(Image.open(img_path)))
+            img_lpng = read_png(img_path)
+            self.assertTrue(torch.all(img_lpng == img_pil))
```
Comment (cpuhrsch) [inline test/test_image.py]: It might be better to use [assetEqual](https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_utils.py#L795) since it checks for matching dtypes, etc. etc. ; unless you have a specific reason not to?

Comment (r-zenine) [reply test/test_image.py]: Good Catch Thanks

Comment (r-zenine) [reply test/test_image.py]: Good Catch Thanks

File: setup.py (hunk 8 above)
Comment (r-zenine) [reply setup.py]: Might explain some failures of the CI. Thanks! 
I'll try it.

File: setup.py (hunk 9 above)
Comment (r-zenine) [reply setup.py]: To be fair, 
I am having a hard time especially with the windows CI and MacOs, because I did not use those systems in years. 
I am trying to get something to work ( even if it's dirty ). Then I'll clean things up for the merge.

File: setup.py (hunk 8 above)
Comment (r-zenine) [reply setup.py]: Thanks for the tip, It helped with the linux_conda based pipeline. Tests are still failling, but not because of timeouts.
I'll keep investigating.

File: setup.py (hunk 9 above)
Comment (r-zenine) [reply setup.py]: Hi, 
I took a look. I'll grab some code. 
Thanks

**Comment by r-zenine** [comment]
HI, 

Another update, 

Still a work in progress. However, we got `macos_wheel` working.

`conda` based setups are not working, tests are failing because of a failure to load the C extension. It might be happening because we are not using the same compiler for everything. Not sure how to fix it. 
I'll investigate this at the end.

`Windows` pipeline are still running. Crossing fingers. 

Going to work.

**Comment by r-zenine** [comment]
Hi There, 

Making progress here! Even though I have to admit, I am way out of my comfort zone. 

The `linux_conda` pipeline are timing out. Any idea how to fix it?
 
I have tried several times to fix the `windows` pipeline without success. The feedback loop is too slow ( having to push the code and wait for the CI to run). Do you have any tips to make the feedback loop go faster? I guess not having a windows machine ready and configured is not helping either. 

The code is not clean by any means. There are tons of commits, I'll clean things up when we get everything up and running and rebase and clean the commits. 

I would appreciate some help with the windows pipeline if anyone can take a look into it. 

Big thanks guys, 
Going to work

**Comment by cpuhrsch** [comment]
Thanks for working on this @r-zenine! Believe me, I understand how frustrating it is to add a dependency. Here are some suggestions on how to break this problem into smaller pieces given that this is becoming too complex.

You could split this into two PRs: png and jpeg separately. That way we can reduce the error surface. In particular if you can split it such that only one dependency needs to be introduced. 

On Windows: I'd focus on fixing all the other cases first. In fact, there is nothing wrong with selectively disabling the support for these ops for Windows and Mac first. You can use [@unittest.skipIf('win' in sys.platform)](https://github.com/pytorch/vision/blob/aadbed6fc0db5b92dc87c2da715adcdedc02e72a/test/test_utils.py#L51) to disable the tests on Windows. We can then later on add support for this.

On the topic of speeding up the feedback loop: Ninja was supposed to help with build times but until we have support for a flag that limits the threads based on the environment we can't enable it just yet. Windows is always its own story and in my experience has terrible feedback loop times. We're working on fixing this, but a good way to avoid it for now is to add support for it in a separate PR.

So, in summary, I'd create a PR that adds jpeg reader support for linux only for now. Then a PR that does the same for png. Then a PR for Mac support for jpeg and png each. Then a PR for Windows support for jpeg and png each.

I hope this helps.

**Comment by r-zenine** [comment]
Hi @cpuhrsch , 

Let's proceed that way. I'll do PR's for: 

1. `png` for `linux`
2. `jpeg` for `linux`
3. `png` for `mac`
4. `jpeg` for `mac`
5. `png` for `windows`
6. `jpeg` for `windows`

In the meantime, if we could find help for the `windows` pipeline, I would greatly appreciate it. 
Thanks

**Comment by cpuhrsch** [comment]
@r-zenine - sounds great! I'll drum up some help once we get to Windows.

**Comment by r-zenine** [comment]
@cpuhrsch, 

Awesome, Thanks !

**Comment by r-zenine** [comment]
Hi @cpuhrsch , @fmassa , 

I have a question for the `jpeg` integration. The code for `libjpeg-turbo` is working correctly and the packaging is ok. However we do not produce exactly the same result as `pillow`. After doing some google. It's seems that it's a normal behavior for Jpeg. 

See here for example : https://github.com/python-pillow/Pillow/issues/3833

Therefore, it makes it difficult to write a non regression test agaist the stock pillow. 

We have two options here : 
Either we build a version of pillow with `turbojpeg` support enabled in the pipeline and we test against it. Or, I use stock `libjpeg` for the code and we drop `turbojpeg`. 

Thoughts?

**Comment by fmassa** [comment]
Subsumed by #2388 and #2382
